sql/service_wsrep.cc

```diff
@@ -183,7 +183,21 @@ extern "C" my_bool wsrep_thd_is_BF(const THD *thd, my_bool sync)
 
 extern "C" my_bool wsrep_thd_is_SR(const THD *thd)
 {
-  return thd && thd->wsrep_cs().transaction().is_streaming();
+  return thd && thd->wsrep_cs().transaction().is_streaming() &&
+    thd->wsrep_cs().transaction().state() == wsrep::transaction::s_executing;
```

Comment:
Can you explain why this change was needed?

Reply:
BF detection logic is only interested in executing SR transactions, as any other state (maybe except prepared) is a transition state, a "technical" state, and in those states SR transaction is no different from any other one-shot transaction (single write set). We don't want to BF abort SR transactions in any other state than executing. The same is true for other parts where SR transactions are checked: during transaction execution, we're only interested if it's a running SR transaction, not in aborting or in prepared or in committing or in any other state.

From this standpoint, I would argue that SR transaction is either executing or otherwise it's semantically equivalent to a single write set transaction. That's why it makes sense to call something an SR transaction only when it's in the executing state.

Reply:
wsrep_thd_is_SR() is also used to decide whether to call wsrep_handle_SR_rollback() during rollback processing. In that context, a transaction is streaming if it has replicated one or more fragments. If that's the case, we want to send a rollback fragment before releasing innodb locks.
Since this is in rollback path (any kind of rollback be it voluntary, bf abort or native deadlock), I'm afraid that the transaction could be in s_must_abort state for example. Failing to send rollback fragment before releasing locks has caused challenging bugs in the past.

Reply:
I checked this code path before, and from what I gather `wsrep_handle_SR_rollback()` solely performs BF abort on an SR transaction, which in turn calls `streaming_rollback()`, but only for SR transaction in executing state.
I also checked more wsrep-lib code to ensure that `streaming_rollback()` is called from only three places:
- `certify_fragment()` during an error path
- `before_rollback()`, which doesn't relate to `wsrep_handle_SR_rollback()` code
- `bf_abort()`, which is mentioned earlier

While I agree that such changes need more judging, I think "catch-all" style coding when it's hard to reason about particular states, and which produces false-positives like erroneous debug messages or attempts to perform actions when they should not be performed, is even worse. I do think that trying to BF abort SR transaction in a wrong state issuing wrong debug messages is bad, especially that it erodes understanding what should be called when and under which invariant.

If this specific case is buggy and produced problems in the past, I would expect it to be covered with MTR tests, but I don't see any failures at the moment.

Reply:
> [...] which in turn calls streaming_rollback(), but only for SR transaction in executing state.
thanks for clarifying this, I missed it when checking the code path